File: 2023/utils/output.ts
export function printGrid(
  grid: any[][],
  mapper: (v: any) => string = (v) => v,
  delay = 0,
) {
  for (let i = 0; i < grid.length; i++) {
    const row = grid[i];
    process.stdout.cursorTo(0, Number(i));
    process.stdout.write(`${row.map(mapper).join(" ")}\n`);
  }
  if (delay) {
    sleep(delay);
  }
}

export function sleep(ms: number) {
  const start = Date.now();
  while (Date.now() - start < ms) {}
}

File: 2019/utils/output.py
import os
import sys
import time

from utils.grid import get_min_max_xy


def print_grid(grid, mapper=lambda v: str(v), delay=0, padding=False):
    # os.system("cls" if os.name == "nt" else "clear")
    for i, row in enumerate(grid):
        sys.stdout.write("\033[K")  # Clear the current line
        sys.stdout.write(
            "\033[{};{}H".format(i + 1, 0)
        )  # Move the cursor to the beginning of the line
        sys.stdout.write((" " if padding else "").join(map(mapper, row)) + "\n")
        sys.stdout.flush()
        # print("".join(map(mapper, row)))

    if delay:
        time.sleep(delay)


def print_grid_dict(grid, mapper=lambda v: str(v), delay=0):
    os.system("cls" if os.name == "nt" else "clear")
    min_x, max_x, min_y, max_y = get_min_max_xy(grid)

    for y in range(min_y, max_y + 1):
        print("".join(mapper((x, y)) for x in range(min_x, max_x + 1)))

    if delay:
        time.sleep(delay)
